compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -1450,7 +1449,8 @@ private void fillEffectiveType(BIntersectionType intersectionType,
         Iterator<BType> iterator = intersectionType.getConstituentTypes().iterator();
         BType effectiveType = iterator.next();
         BLangType bLangEffectiveType = bLangTypeItr.next();
-        if (effectiveType.tag == TypeTags.READONLY && iterator.hasNext()) {
+        BType effectiveImpliedTypeReferenceType = Types.getImpliedType(effectiveType);
+        if (effectiveImpliedTypeReferenceType.tag == TypeTags.READONLY && iterator.hasNext()) {
```

Comment:
Can't we just use impliedType? Not sure if `bLangEffectiveType` is accurate to begin with.

Reply:
updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42025/commits/ff5949d36f10f43f3f0843ab7a88999874407333).